src/input.rs

```diff
@@ -130,9 +132,14 @@ impl<'a> Input<'a> {
 
     fn _ordinary_file(path: &Path) -> Self {
         let kind = InputKind::OrdinaryFile(path.to_path_buf());
+        let metadata = InputMetadata {
+            size: fs::metadata(path).map(|m| m.len()).ok(),
```

Comment:
It's probably irrelevant since other things (like actually reading the file) are much slower, but this is an additional (stat) syscall for every file that we print - *even if we do not print the filesize*. It might make sense to run a quick benchmark in a scenario where we print a lot of small files.

If it has a significant performance impact, we could delay loading the metadata until we actually need it.

Reply:
I ran a benchmark against 100 small files.

master branch
```
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `bat … --language=txt *.txt` | 46.6 ± 2.2 | 43.6 | 55.2 | 1.00 |

```

My branch:
```
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `bat … --language=txt *.txt` | 47.3 ± 2.0 | 44.3 | 57.0 | 1.00 |
```

But while running these benchmarks, I noticed a large regression in performance between 0.18.3 and master branch on "Syntax highlighting speed jquery" (wrap=character and wrap=never). From milliseconds to 14s, it's definitely worth investigating:

master branch
```
       │ File: bat-master/syntax-highlighting-speed-jquery.js.md
───────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ | Command | Mean [s] | Min [s] | Max [s] | Relative |
   2   │ |:---|---:|---:|---:|---:|
   3   │ | `bat … jquery.js` | 13.258 ± 0.070 | 13.174 | 13.392 | 1.00 |

```

0.18.3
```
       │ File: bat-system/syntax-highlighting-speed-jquery.js.md
───────┼──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ | Command | Mean [s] | Min [s] | Max [s] | Relative |
   2   │ |:---|---:|---:|---:|---:|
   3   │ | `bat … jquery.js` | 1.052 ± 0.008 | 1.039 | 1.062 | 1.00 |
```

Reply:
I'm not able to reproduce that big regression. Are you sure you used release builds in both cases?

`hyperfine --export-markdown=/dev/stdout -L bat bat-master,bat-v0.18.3 '{bat} --no-config --style=full --color=always --wrap=never --terminal-width=80 tests/benchmarks/highlighting-speed-src/jquery.js'`

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `bat-master --no-config --style=full --color=always --wrap=never --terminal-width=80 tests/benchmarks/highlighting-speed-src/jquery.js` | 366.2 ± 1.5 | 364.8 | 369.5 | 1.02 ± 0.00 |
| `bat-v0.18.3 --no-config --style=full --color=always --wrap=never --terminal-width=80 tests/benchmarks/highlighting-speed-src/jquery.js` | 357.9 ± 0.7 | 357.2 | 359.1 | 1.00 |

The difference is even less pronounced with `--wrap=character`.

Reply:
@Enselic ah you are right, I was using debug builds instead of release builds. 🤦 😁 

New runs with release builds for many small files:

my branch:
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `bat … --language=txt *.txt` | 12.8 ± 1.4 | 10.8 | 20.7 | 1.00 |


1.8.3
| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `bat … --language=txt *.txt` | 13.1 ± 1.4 | 11.2 | 24.2 | 1.00 |

And I don't see any significant difference between syntax highlighting benchmarks either, so all good 👍

Reply:
Thanks for adding a new performance benchmark!

I've also compared master vs this PR using all existing and the new performance benchmarks, and the difference is negligible (meaning: I don't really see any difference). I also used a low-end desktop for this, so if we had a problem, I should have seen it.

So this is a non-issue in practice as far as I can tell. It was good to make sure that was the case though!

Besides, if this turns out to be a problem later for whatever reason, it should be no big deal to change it to be lazy-loaded. It's an implementation detail that should not require changing public API.